src/google/protobuf/descriptor.proto

```diff
@@ -481,7 +490,11 @@ message FieldOptions {
   // For Google-internal migration only. Do not use.
   optional bool weak = 10 [default=false];
 
-
+  // Provides the ability to override the name of the property
+  // generated for this field. This is applied to all properties
+  // and methods to do with this field, including HasFoo, FooCount,
+  // FooList etc.
+  optional string csharp_property_name = 11;
```

Comment:
Simplest example of when it would happen is if you have a field with the same name as the message (barring capitalization). For example:

```
message Name {
  string name = 1;
}
```

(It used to be that "Foo foo" would cause a problem, where "Foo" was a nested message, which is fairly common, but the addition of the "Types" nested class between a parent class and a nested message fixed that... at the cost of simplicity.)